lang/py/src/avro/schema.py

```diff
@@ -166,68 +170,62 @@ def to_json(self, names):
 class Name(object):
   """Class to describe Avro name."""
 
+  _base_name_pattern = re.compile(r'(?:^|\.)[A-Za-z_][A-Za-z0-9_]*$')
+
   def __init__(self, name_attr, space_attr, default_space):
     """
     Formulate full name according to the specification.
 
     @arg name_attr: name value read in schema or None.
     @arg space_attr: namespace value read in schema or None.
-    @ard default_space: the current default space or None.
+    @arg default_space: the current default space or None.
     """
     # Ensure valid ctor args
-    if not (isinstance(name_attr, basestring) or (name_attr is None)):
-      fail_msg = 'Name must be non-empty string or None.'
-      raise SchemaParseException(fail_msg)
-    elif name_attr == "":
-      fail_msg = 'Name must be non-empty string or None.'
+    fail_msg = 'Name must be non-empty string or None.'
+    if name_attr == "" or not (isinstance(name_attr, basestring) or (name_attr is None)):
       raise SchemaParseException(fail_msg)
 
-    if not (isinstance(space_attr, basestring) or (space_attr is None)):
-      fail_msg = 'Space must be non-empty string or None.'
-      raise SchemaParseException(fail_msg)
-    elif name_attr == "":
-      fail_msg = 'Space must be non-empty string or None.'
-      raise SchemaParseException(fail_msg)
+    self._full = None
 
-    if not (isinstance(default_space, basestring) or (default_space is None)):
-      fail_msg = 'Default space must be non-empty string or None.'
-      raise SchemaParseException(fail_msg)
-    elif name_attr == "":
-      fail_msg = 'Default must be non-empty string or None.'
-      raise SchemaParseException(fail_msg)
+    if not name_attr:
+      return
 
-    self._full = None;
+    if not self._base_name_pattern.search(name_attr):
+      raise InvalidName("{!s} is not a valid Avro name because it "
+                        "does not match the pattern {!s}".format(
+                          name_attr, self._base_name_pattern.pattern))
 
-    if name_attr is None or name_attr == "":
-        return;
 
-    if (name_attr.find('.') < 0):
-      if (space_attr is not None) and (space_attr != ""):
-        self._full = "%s.%s" % (space_attr, name_attr)
-      else:
-        if (default_space is not None) and (default_space != ""):
-           self._full = "%s.%s" % (default_space, name_attr)
-        else:
-          self._full = name_attr
+    fail_msg = 'Namespace must be non-empty string or None.'
```

Comment:
This might be an existing bug: [The empty string may also be used as a namespace to indicate the null namespace.](https://avro.apache.org/docs/1.9.1/spec.html#names).

In other words, if the namespace is "", the default namespace should not be applied `schema.Name('a', '', 'c.b').fullname` should be `'a'`